### PR TITLE
feat(web-scripts): add typechecking to lint task

### PR DESCRIPTION
**Pull request by fastfrwrd**
We've been talking about making a few changes to the base behavior of `web-scripts lint`:

1. We would like there to be a build artifact output by ESLint by default. This would help us to assess whether builds ran static analysis without having to do invasive checks.
2. We would like to run typecheks by default as part of linting. This will help us move to using Babel to build if we decide that makes sense eventually; it will also help us avoid needing to run `web-scripts build` to properly test a project, which is a bit unintuitive at the moment.

This PR:
- **adds a type check by default to `web-scripts lint`.** You can turn it off by using `--no-typecheck`.
- **adds a parallel ESLint run which outputs Checkstyle**. Since you can't use two formatters for ESLint, and ESLint forces you to write files yourself using stdout, this seems to be the only option. You can turn this off with `--no-output`.

### CODE REVIEW

File: packages/web-scripts/src/allSettled.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            },
+            (reason: Error) => {
+              const result: PromiseRejection = { status: 'rejected', reason };
+              resolve(result);
+            },
+          );
+        }),
+    ),
+  );
+}
```
Comment (fastfrwrd) [inline packages/web-scripts/src/allSettled.ts]: I tried to use https://www.npmjs.com/package/promise.allsettled but it was not working very well for me. this is a simple implementation of https://github.com/tc39/proposal-promise-allSettled, which is in stage 4 but doesn't have wide adoption yet it seems.

File: packages/web-scripts/src/allSettled.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+interface PromiseRejection extends PromiseResult {
+  status: 'rejected';
+  reason: Error;
+}
+interface PromiseFulfillment<T> extends PromiseResult {
+  status: 'fulfilled';
+  value: T;
+}
+
+export function allSettled<T>(arr: Promise<T>[]): Promise<PromiseResult[]> {
```
Comment (kirbysayshi) [inline packages/web-scripts/src/allSettled.ts]: Another, possibly more idiomatic, way to accomplish these types is:

```ts
type PromiseRejection {
  status: 'rejected';
  reason: Error;
}
type PromiseFulfillment<T> {
  status: 'fulfilled';
  value: T;
}

type PromiseResult = PromiseRejection | PromiseFulfillment;
```

Later, when you use this function, you can even use a `switch` statement or `if` and get autocomplete for valid `status`!

Comment (fastfrwrd) [reply packages/web-scripts/src/allSettled.ts]: YES this is better.

Comment (fastfrwrd) [reply packages/web-scripts/src/allSettled.ts]: the `allSettled` module ultimately did not end up being necessary, but I'm glad I learned this.

File: packages/web-scripts/src/Tasks/LintTask.ts [hunk 3]
```diff
@@ -21,8 +51,30 @@ export function lintTask(task: LintTaskDesc): SpawnSyncReturns<Buffer> {
...
+  const stdout = await spawn(cmd, args, { stdio: 'inherit', ...options });
+  return (stdout || '').toString();
+}
+
+async function eslintRunWithCheckstyle(task: LintTaskDesc): Promise<string> {
+  return await eslintRun(
+    {
+      ...task,
+      format: 'checkstyle',
+      outputFile: 'checkstyle-result.xml',
```
Comment (fastfrwrd) [inline packages/web-scripts/src/Tasks/LintTask.ts]: outputFile guarantees that the file is written, even if things are throwing. That means that we don't need `allSettled`.

**Comment by fastfrwrd** [comment]
@kirbysayshi I've been wondering if the `jest` run ought to be moved to `pre-push` by default. I've seen a few open source projects using that lately. It would mean a new task and updates to repos, but I feel it'd be worth it if we're feeling pain there now.

**Comment by fastfrwrd** [comment]
I made some changes to this after we talked a bit on the team.

- It makes a lot more sense and is more backwards compatible if we just make users run lint twice in parallel to get the checkstyle output. It should be pretty easy to do that internally here.
- I've marked this as a breaking change, because it is possible that it will break people's builds.

**Comment by kirbysayshi** [comment]
@fastfrwrd if anything, I'd prefer fewer things running based on git hooks. Pushing, committing, rebasing, all those things should be as fast as possible!

**Pull request merged**